File: leetcode/FlattenBinaryTree.java
public class FlattenBinaryTree {
    public static void main(String[] args) {
        TreeNode t1 = new TreeNode(1);
        TreeNode t2 = new TreeNode(2);
        TreeNode t3 = new TreeNode(3);
        TreeNode t4 = new TreeNode(4);
        TreeNode t5 = new TreeNode(5);
        TreeNode t6 = new TreeNode(6);
        TreeNode t7 = new TreeNode(7);
        TreeNode t8 = new TreeNode(8);
        TreeNode t9 = new TreeNode(9);

        t1.left = t2;
        t2.left = t3;
        t3.right = t4;
        t2.right = t5;
        t1.right = t6;
        t6.left = t7;
        t6.right = t8;
        t8.right = t9;

        printTree(t1, 0);
        flatten(t1);
        printTree(t1, 0);
    }
    public static void printTree(TreeNode root, int depth)
    {
        for (int i = 0; i < depth; i++)
            System.out.print("    ");
        if (root == null)
        {
            System.out.println("-");
            return;
        }
        else
        {
            System.out.println(root.val);
            printTree(root.left, depth + 1);
            printTree(root.right, depth + 1);
        }
    }

    public static void flatten(TreeNode root)
    {
        if (root == null)
            return;

        flatten(root.right);

        if (root.left == null)
            return;

        flatten(root.left);
        endOfList(root.left).right = root.right;
        root.right = root.left;
        root.left = null;
    }
    public static TreeNode endOfList(TreeNode root)
    {
        while (root.right != null)
        {
            root = root.right;
        }
        return root;
    }
}

class TreeNode
{
    int val;
    TreeNode left;
    TreeNode right;
    public TreeNode(int x)
    {
        this.val = x;
    }
}

File: leetcode/114/FlattenBinaryTree.py
import sys

def main():
    input_list = eval(sys.argv[1])
    root = parseTreeFromList(input_list)

    printTree(root, 0)
    flatten(root)
    printDoublyLinkedList(root)

def flatten(root):
    oldLeft, oldRight = root.left, root.right
    root.left, root.right = None, None

    if oldLeft != None:
        flatten(oldLeft)
        root.right = oldLeft
        # root.right.left = root # uncomment this line for doubly-linked list!
    if (oldRight != None):
        end = endOfList(root)
        flatten(oldRight)
        end.right = oldRight
        # oldRight.left = end # uncomment this line for doubly-linked list!

def endOfList(head):
    while head.right != None:
        head = head.right
    return head


# TESTING INFRASTRUCTURE
# ==============================

class TreeNode():
    def __init__(self, val, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

def parseTreeFromList(l):
    nodes = [TreeNode(l[i]) for i in range(len(l))]
    for i in range(len(nodes)):
        if i*2 + 1 < len(nodes):
            nodes[i].left = nodes[i*2 + 1]
        if i*2 + 2 < len(nodes):
            nodes[i].right = nodes[i*2 + 2]
    if nodes != []:
        return nodes[0]
    else:
        raise Error("need non-empty python list as argument")

def printTree(node, depth):
    indent = "    " * depth
    val = "-" if node == None else node.val
    print(indent + str(val))

    if (node != None):
        printTree(node.left, depth+1)
        printTree(node.right, depth+1)

def printDoublyLinkedList(head):
    while head != None:
        left = "None" if head.left == None else str(head.left.val)
        left = left.ljust(5)
        val = head.val
        right = "None" if head.right == None else str(head.right.val)
        right = right.rjust(5)
        print("{} <-- {} --> {}".format(left, val, right))

        head = head.right


if __name__ == "__main__":
    main()
